pkg/roachprod/blobfixture/registry.go

```diff
@@ -0,0 +1,282 @@
+}
+
+// GC deletes fixtures that were leaked or obsolete. See the comment on the
+// fixturesToGc for details about the GC policy.
+func (r *Registry) GC(ctx context.Context, l *logger.Logger) error {
+	// List objects in /roachtest/v25.1/metadata
+	fixtures, err := r.listFixtures(ctx, path.Join(r.uri.Path, "metadata"), l)
+	if err != nil {
+		return errors.Wrap(err, "gc failed to list metadata")
+	}
+
+	toDelete := fixturesToGc(r.clock(), fixtures)
+
+	l.Printf("running gc: deleting %d out of %d fixtures", len(toDelete), len(fixtures))
+
+	for _, f := range toDelete {
+		l.Printf("deleting fixture %q: %s", f.metadata.DataPath, f.reason)
+		if err := r.deleteBlobsMatchingPrefix(f.metadata.DataPath); err != nil {
+			return err
+		}
+		if err := r.deleteMetadata(f.metadata); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *Registry) Close() {
+	_ = r.storage.Close()
+}
+
+// URI returns a new URL with the given path appended to the registry's URI.
+//
+// Example:
+// fixture = r.GetLatest(ctx, "backup-tpcc-30k")
+// r.URI(fixture.DataPath) returns 'gs://cockroach-fixtures/roachprod/v25.1/backup-tpcc-30k/20220101-1504'
+func (r *Registry) URI(path string) url.URL {
+	copy := r.uri
+	copy.Path = path
+	return copy
+}
+
+func (r *Registry) listFixtures(
+	ctx context.Context, kindPrefix string, l *logger.Logger,
+) ([]FixtureMetadata, error) {
+	if l != nil {
+		l.Printf("listing fixtures: %s", kindPrefix)
+	}
+	result := []FixtureMetadata{}
+	// List objects in prefix
+	err := r.storage.List(ctx, kindPrefix /*delimiter*/, "", func(found string) error {
+		// TODO(jeffswenson): does this need to change behavior base on cloud vs
+		// local?
+		file, _, err := r.storage.ReadFile(ctx, path.Join(kindPrefix, found), cloud.ReadOptions{})
+		if err != nil {
+			return err
+		}
+		defer func() { _ = file.Close(ctx) }()
+
+		json, err := ioctx.ReadAll(ctx, file)
+		if err != nil {
+			return err
+		}
+
+		metadata := FixtureMetadata{}
+		if err := metadata.UnmarshalJson(json); err != nil {
+			return err
+		}
+
+		result = append(result, metadata)
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func (r *Registry) upsertMetadata(metadata FixtureMetadata) error {
+	json, err := metadata.MarshalJson()
+	if err != nil {
+		return err
+	}
+
+	writer, err := r.storage.Writer(context.Background(), metadata.MetadataPath)
+	if err != nil {
+		return err
+	}
+
+	if _, err := writer.Write(json); err != nil {
+		_ = writer.Close()
+		return err
+	}
+
+	return writer.Close()
+}
+
+func (r *Registry) deleteMetadata(metadata FixtureMetadata) error {
+	return errors.Wrap(r.storage.Delete(context.Background(), metadata.MetadataPath), "unable to delete fixture metadata")
+}
+
+func (r *Registry) deleteBlobsMatchingPrefix(prefix string) error {
+	return r.storage.List(context.Background(), prefix, "", func(path string) error {
+		return r.storage.Delete(context.Background(), prefix+path)
```

Comment:
i hope this scales!

Reply:
The list is going to be fast. We may need to parallelize the deletes, but I'm trying to write the simplest possible code on the first pass.